android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboardDownloaderActivity.java

```diff
@@ -0,0 +1,446 @@
+      if (filename == null || filename.isEmpty()) {
+        filename = "unknown";
+      }
+    } else {
+      return;
+    }
+
+    Bundle args = new Bundle();
+    String title = "";
+    if (url != null) {
+      title = "Custom Keyboard: " + filename;
+    } else if (customKeyboard != null && customLanguage != null &&
+        !customKeyboard.trim().isEmpty() && !customLanguage.trim().isEmpty()) {
+      int kbIndex = KMManager.getKeyboardIndex(getApplicationContext(), customKeyboard, customLanguage);
+
+      if (kbIndex >= 0) {
+        KMManager.setKeyboard(getApplicationContext(), kbIndex);
+        // No interaction needed
+        return; // false
+      }
+
+      title = customLanguage + "_" + customKeyboard;
+  } else {
+    // Download keyboard from cloud server
+    title = langName + ": " + kbName;
+  }
+
+  DialogFragment dialog = new ConfirmDialogFragment();
+    args.putString(ConfirmDialogFragment.ARG_TITLE, title);
+    dialog.setArguments(args);
+    dialog.show(getFragmentManager(), "dialog");
+}
+
+  /**
+   * Async task to download a Keyman keyboard from either Keyman cloud server or custom url
+   * @param context
+   * @param showProgressDialog
+   */
+    public static void download(final Context context, final boolean showProgressDialog) {
+
+    new AsyncTask<Void, Integer, Integer>() {
+      private ProgressDialog progressDialog;
+      //private String languageName = "";
+      //private String keyboardName = "";
+      private String kbVersion = "1.0";
+      private String kbIsCustom = isCustom ? "Y" : "N";
+      private String font = "";
+      private String oskFont = "";
+
+      @Override
+      protected void onPreExecute() {
+        super.onPreExecute();
+        if (showProgressDialog) {
+          progressDialog = new ProgressDialog(context);
+          progressDialog.setMessage("Downloading keyboard...");
+          progressDialog.setCancelable(false);
+          if (!((Activity) context).isFinishing()) {
+            progressDialog.show();
+          } else {
+            cancel(true);
+            progressDialog = null;
+          }
+        }
+      }
+
+      @Override
+      protected Integer doInBackground(Void... voids) {
+        int ret = -1;
+        ArrayList<String> urls = new ArrayList<String>();
+
+        if (isCancelled())
+          return ret;
+
+        try {
+          String exceptionStr = "Invalid keyboard";
+          if (pkgID == null || pkgID.trim().isEmpty() ||
+            (!isCustom && (langID == null || langID.trim().isEmpty() || kbID == null || kbID.trim().isEmpty()))) {
+            throw new Exception(exceptionStr);
+          }
+
+          String deviceType = context.getResources().getString(R.string.device_type);
+          if (deviceType.equals("AndroidTablet")) {
+            deviceType = "androidtablet";
+          } else {
+            deviceType = "androidphone";
+          }
+
+          // Placeholder here to attempt to process .kmp keyboard first:
```

Comment:
I take it this is a placeholder for when we get the .kmp APIs?

Reply:
yeah, I think we'd prefer getting keyboards via:
1. KMP from Keyman cloud
2. ad-hoc KMP

3. current delivery from Keyman cloud
4. ad-hoc json